### PR TITLE
Azure: Added support for multiple VMSS communication 

### PR DESCRIPTION
**Pull request by purushothamank**
- To support Blue-Green Deployment model
- Added new option vmssList (supply mutiple vmss names)

### CODE REVIEW

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: azure.js [hunk 1]
```diff
@@ -62,19 +62,77 @@ module.exports = (args, sendTo) => {
...
+  ) {
+    return new Promise((resolve, reject) => {
+      client.networkInterfaces._listVirtualMachineScaleSetNetworkInterfaces(
+        args.azure,
+        vmssName,
+        function (err, results) {
+          if (err) {
+            // Do Nothing if the VMSS is deleted, so than for next deployment inter VMSS forwarding will work
+            if (
+              err.message !==
```
[the remark this answers is not in the document]
Comment (purushothamank) [reply azure.js]: Updated

File: azure.js [hunk 2]
```diff
@@ -62,19 +62,77 @@ module.exports = (args, sendTo) => {
...
+    vmssName,
+    vmssNameList
+  ) {
+    return new Promise((resolve, reject) => {
+      client.networkInterfaces._listVirtualMachineScaleSetNetworkInterfaces(
+        args.azure,
+        vmssName,
+        function (err, results) {
+          if (err) {
+            // Do Nothing if the VMSS is deleted, so than for next deployment inter VMSS forwarding will work
```
[the remark this answers is not in the document]
Comment (purushothamank) [reply azure.js]: Updated.

File: package.json [hunk 3]
```diff
@@ -1,15 +1,15 @@
 {
   "name": "repeater",
-  "version": "0.1.0",
+  "version": "0.1.1",
   "description": "Multicast Datagram Repeater",
   "main": "repeater.js",
   "bin": "./repeater.js",
   "dependencies": {
+    "azure-arm-network": "^11.1.0",
```
[the remark this answers is not in the document]
Comment (purushothamank) [reply package.json:8]: With latest package, _listVirtualMachineScaleSetNetworkInterfaces method is not visible.
which breaks our implementation due to which we need to continue with azure-arm-network for now.

File: repeater.js [hunk 4]
```diff
@@ -34,6 +34,11 @@ const args = require('yargs')
     type: 'string',
     requiresArg: false
   })
+  .option('vmssList', {
+    describe: 'Get the list of names of required VMSS Scale Sets',
+    type: 'array',
+    requiresArg: false
+  })
```
Comment (mitza-oci) [inline repeater.js:41]: Requiring the names of all VMSSs to be known at startup is too limiting -- at the time the repeater is started the name of another VMSS isn't known.  

Instead could this use Azure's tags?  I assume there's a way to use their API to find resources (in this case VMSSs) by tag.

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: azure.js [hunk 5]
```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
...
   const updateFunc = () => {
     MsRest.loginWithVmMSI().then(credentials => {
       let client = new NetworkManagementClient(credentials, subscriptionId)
       const ips = []
 
-      if ( typeof args.vmss !== 'undefined' && args.vmss ) {
+      if (typeof args.vmss !== 'undefined' && args.vmss) {
         processVirtualMachineScaleSet(client, ips, getLocalIps(), args.vmss)
+      } else if (typeof args.vmssTag !== 'undefined' && args.vmssTag) {
+        console.log(args.vmssTag);
```
Comment (mitza-oci) [inline azure.js:150]: Remove logging? (if it's just for dev testing)

File: azure.js [hunk 6]
```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
...
+          const results = await processVirtualMachineScaleSetPromise(client, vmssName)
+          if (results && results.length) {
+            results.forEach(result => {
+              result.ipConfigurations.forEach(ip => {
+                if (!localIps.includes(ip.privateIPAddress)) {
+                  ips.push(ip.privateIPAddress)
+                }
+              })
+            })
+            resolve(ips)
```
Comment (mitza-oci) [inline azure.js:129]: After this line I think there should be a `return` or an `else` so control doesn't reach `resolve([])` 2 lines below

File: azure.js [hunk 7]
```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
...
 
+  async function getVMDetailsByTag(tagKey, tagValue) {
+
+    const credential = new DefaultAzureCredential();
+
+    const client = new ComputeManagementClient(credential, subscriptionId);
+
+    // Get the list of VMs
+    //const vms = await client.virtualMachines.listAll().byPage().next();
+
```
Comment (mitza-oci) [inline azure.js:99]: Remove these lines?

File: azure.js [hunk 8]
```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
...
+    const vmss = await client.virtualMachineScaleSets.listAll().byPage().next();
+
+    const filteredVMSSs = vmss.value
+      .filter(vms => vms.tags && vms.tags[tagKey] === tagValue)
+      .map(vms => vms.name);
+
+    return filteredVMSSs;
+
```
Comment (mitza-oci) [inline azure.js:107]:
```suggestion
    return vmss.value
      .filter(vms => vms.tags && vms.tags[tagKey] === tagValue)
      .map(vms => vms.name);
```

File: azure.js [hunk 9]
```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
...
+    client,
+    vmssName
+  ) {
+    return new Promise((resolve, reject) => {
+      client.networkInterfaces._listVirtualMachineScaleSetNetworkInterfaces(
+        args.azure,
+        vmssName,
+        function (err, results) {
+          if (err) {
+            // Do nothing if the VMSS is deleted, so that for next deployment inter VMSS forwarding will work
```
Comment (mitza-oci) [inline azure.js:79]: Is this still needed when using tags?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: azure.js (hunk 6 above)
Comment (purushothamank) [reply azure.js:124]: Return is not required here we are using outer scoped variable ips, we need this promise just for retrieving all VM's IPs at once.

File: azure.js (hunk 9 above)
Comment (purushothamank) [reply azure.js:79]: Here we are just reducing the error logs even with tags we may get this error once during the vmss delete.

File: azure.js (hunk 6 above)
Comment (mitza-oci) [reply azure.js:124]: After the first resolve() is called, the second call doesn't do anything.

**Review by mitza-oci — APPROVED** [review]

File: azure.js (hunk 6 above)
Comment (purushothamank) [reply azure.js]: Resolved

**Pull request merged**